Objects/dictobject.c

```diff
@@ -875,69 +875,37 @@ lookdict_index(PyDictKeysObject *k, Py_hash_t hash, Py_ssize_t index)
     Py_UNREACHABLE();
 }
 
-// Search non-Unicode key from Unicode table
-static Py_ssize_t
-unicodekeys_lookup_generic(PyDictObject *mp, PyDictKeysObject* dk, PyObject *key, Py_hash_t hash)
+typedef void *(*generic_entries_func)(PyDictKeysObject *dk);
+
+static inline Py_ALWAYS_INLINE Py_ssize_t
+do_lookup(PyDictObject *mp, PyDictKeysObject *dk, PyObject *key, Py_hash_t hash,
+          generic_entries_func entries,
+          Py_ssize_t (*check_lookup)(PyDictObject *, PyDictKeysObject *, void *, Py_ssize_t ix, PyObject *key, Py_hash_t))
 {
-    PyDictUnicodeEntry *ep0 = DK_UNICODE_ENTRIES(dk);
+    void *ep0 = entries(dk);
     size_t mask = DK_MASK(dk);
     size_t perturb = hash;
     size_t i = (size_t)hash & mask;
     Py_ssize_t ix;
     for (;;) {
         ix = dictkeys_get_index(dk, i);
         if (ix >= 0) {
-            PyDictUnicodeEntry *ep = &ep0[ix];
-            assert(ep->me_key != NULL);
-            assert(PyUnicode_CheckExact(ep->me_key));
-            if (ep->me_key == key) {
+            ix = check_lookup(mp, dk, ep0, ix, key, hash);
```

Comment:
I think it would be a bit easier to understand if `check_lookup` followed the convention of returning `0` or `1` for equality and a negative value on error. I was a bit confused at first by the use of `DKIX_DUMMY` for "not equals".

I expect it will also lead to better code generation in some cases too.

Reply:
Good suggestion, I can continue to use the `DKIX_` error codes as there's multiple of them that need to be propagated out and use 0/1 for successful comparisons.